File: tests/resources/set_contract.ml
open Optitrust
open Syntax
open Resource_contract

let _ = Run.script_cpp (fun () ->
    !! Resources.set_fun_contract (parse_fun_contract [__modifies("a ~> Cell")]) [multi cFunDef ["incr"; "incr_twice"]];

    !! Resources.set_fun_contract (parse_fun_contract [__modifies("n ~> Cell"); __modifies("m ~> Cell")]) [cFunDef "incr_both"];

    (* Should produce empty diff: *)
    !! Resources.set_fun_contract (parse_fun_contract [__modifies("n ~> Cell, m ~> Cell")]) [cFunDef "incr_both"];

    !! Resources.set_fun_contract (parse_fun_contract [__modifies("m ~> Matrix1(sz)")]) [cFunDef "incr_range"];
    !! Resources.set_loop_contract (parse_loop_contract [__modifies("&m[MINDEX1(sz, i)] ~> Cell")]) [cFor "i"];

    (*!! Trace.apply Scope.infer_var_ids;
    !! Trace.apply Scope.unique_alpha_rename;*)

    Resources.show ();
)

File: tests/resources/set_contract.cpp
#include "../../include/optitrust.h"

void incr(int* a) {
    __admitted();
    *a = *a + 1;
}

void incr_twice(int* a) {
    incr(a);
    incr(a);
}

void incr_both(int* n, int* m) {
    incr(n);
    incr(m);
}

void incr_range(int* m, int sz) {
    for (int i = 0; i < sz; ++i) {
        incr(&m[MINDEX1(sz, i)]);
    }
}

int main() {}
